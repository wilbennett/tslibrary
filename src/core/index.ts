export * from "./core-types";
export * from "./mathex";
export * from "./promiseex";
