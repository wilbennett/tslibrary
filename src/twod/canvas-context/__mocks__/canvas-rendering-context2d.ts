import { Style } from '..';

export class CanvasRenderingContext2D {
    //================================================================================================================
    // Canvas State
    //================================================================================================================
    save() {
    }

    restore() {
    }

    //================================================================================================================
    // CanvasTransform
    //================================================================================================================
    transformation: number[] = [];
    // getTransform(): DOMMatrix;
    // resetTransform(): void;
    // rotate(angle: number): void;
    // scale(x: number, y: number): void;
    setTransform(a: number, b: number, c: number, d: number, e: number, f: number): void {
        this.transformation = [a, b, c, d, e, f];
    }
    // setTransform(transform?: DOMMatrix2DInit): void;
    // transform(a: number, b: number, c: number, d: number, e: number, f: number): void;
    // translate(x: number, y: number): void;
    //================================================================================================================
    // CanvasCompositing
    //================================================================================================================
    globalAlpha: number = 1;
    globalCompositeOperation: string = "source-over";
    //================================================================================================================
    // CanvasImageSmoothing
    //================================================================================================================
    imageSmoothingEnabled: boolean = false;
    imageSmoothingQuality: ImageSmoothingQuality = "low";
    //================================================================================================================
    // CanvasFillStrokeStyles
    //================================================================================================================
    fillStyle: Style = "#000000";
    strokeStyle: Style = "#000000";

    // @ts-ignore - unused param.
    createLinearGradient(x0: number, y0: number, x1: number, y1: number): CanvasGradient | null {
        return null;
    }

    // @ts-ignore - unused param.
    createPattern(image: CanvasImageSource, repetition: string): CanvasPattern | null { return null; }

    // @ts-ignore - unused param.
    createRadialGradient(x0: number, y0: number, r0: number, x1: number, y1: number, r1: number): CanvasGradient | null {
        return null;
    }
    //================================================================================================================
    // CanvasShadowStyles
    //================================================================================================================
    shadowBlur: number = 0;
    shadowColor: string = "#000000";
    shadowOffsetX: number = 0;
    shadowOffsetY: number = 0;
    //================================================================================================================
    // CanvasFilters
    //================================================================================================================
    // filter: string; //* Proposal.
    //================================================================================================================
    // CanvasRect
    //================================================================================================================
    // @ts-ignore - unused param.
    clearRect(x: number, y: number, w: number, h: number): void { }
    // @ts-ignore - unused param.
    fillRect(x: number, y: number, w: number, h: number): void { }
    // @ts-ignore - unused param.
    strokeRect(x: number, y: number, w: number, h: number): void { }
    //================================================================================================================
    // CanvasDrawPath
    //================================================================================================================
    // beginPath(): void;
    // clip(fillRule?: CanvasFillRule): void;
    // clip(path: Path2D, fillRule?: CanvasFillRule): void;
    // fill(fillRule?: CanvasFillRule): void;
    // fill(path: Path2D, fillRule?: CanvasFillRule): void;
    // isPointInPath(x: number, y: number, fillRule?: CanvasFillRule): boolean;
    // isPointInPath(path: Path2D, x: number, y: number, fillRule?: CanvasFillRule): boolean;
    // isPointInStroke(x: number, y: number): boolean;
    // isPointInStroke(path: Path2D, x: number, y: number): boolean;
    // stroke(): void;
    // stroke(path: Path2D): void;
    //================================================================================================================
    // CanvasUserInterface
    //================================================================================================================
    // drawFocusIfNeeded(element: Element): void;
    // drawFocusIfNeeded(path: Path2D, element: Element): void;
    // scrollPathIntoView(): void;
    // scrollPathIntoView(path: Path2D): void;
    //================================================================================================================
    // CanvasText
    //================================================================================================================
    // fillText(text: string, x: number, y: number, maxWidth?: number): void;
    // measureText(text: string): TextMetrics;
    // strokeText(text: string, x: number, y: number, maxWidth?: number): void;
    //================================================================================================================
    // CanvasDrawImage
    //================================================================================================================
    // drawImage(image: CanvasImageSource, dx: number, dy: number): void;
    // drawImage(image: CanvasImageSource, dx: number, dy: number, dw: number, dh: number): void;
    // drawImage(image: CanvasImageSource, sx: number, sy: number, sw: number, sh: number, dx: number, dy: number, dw: number, dh: number): void;
    //================================================================================================================
    // CanvasImageData
    //================================================================================================================
    // createImageData(sw: number, sh: number): ImageData;
    // createImageData(imagedata: ImageData): ImageData;
    // getImageData(sx: number, sy: number, sw: number, sh: number): ImageData;
    // putImageData(imagedata: ImageData, dx: number, dy: number): void;
    // putImageData(imagedata: ImageData, dx: number, dy: number, dirtyX: number, dirtyY: number, dirtyWidth: number, dirtyHeight: number): void;
    //================================================================================================================
    // CanvasPathDrawingStyles
    //================================================================================================================
    lineCap: CanvasLineCap = "butt";
    lineDashOffset: number = 0;
    lineJoin: CanvasLineJoin = "miter";
    lineWidth: number = 1;
    miterLimit: number = 10;
    // lineDash: number[] = [];
    // getLineDash(): number[];
    // setLineDash(segments: number[]): void;
    //================================================================================================================
    // CanvasTextDrawingStyles
    //================================================================================================================
    font: string = "10px sans-serif";
    textAlign: CanvasTextAlign = "start";
    textBaseline: CanvasTextBaseline = "alphabetic";
    // direction: CanvasDirection; //* Proposal.
    //================================================================================================================
    // CanvasPath
    //================================================================================================================
    // arc(x: number, y: number, radius: number, startAngle: number, endAngle: number, anticlockwise?: boolean): void;
    // arcTo(x1: number, y1: number, x2: number, y2: number, radius: number): void;
    // bezierCurveTo(cp1x: number, cp1y: number, cp2x: number, cp2y: number, x: number, y: number): void;
    // closePath(): void;
    // ellipse(x: number, y: number, radiusX: number, radiusY: number, rotation: number, startAngle: number, endAngle: number, anticlockwise?: boolean): void;
    // lineTo(x: number, y: number): void;
    // moveTo(x: number, y: number): void;
    // quadraticCurveTo(cpx: number, cpy: number, x: number, y: number): void;
    // rect(x: number, y: number, w: number, h: number): void;
}
