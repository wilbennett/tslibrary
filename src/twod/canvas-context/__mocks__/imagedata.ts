export class ImageData {
    constructor(public width: number, public height: number) {
    }

    data!: Uint8ClampedArray;
}
