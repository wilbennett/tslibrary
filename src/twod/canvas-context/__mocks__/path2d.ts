export class Path2D {
    // @ts-ignore - unused param.
    addPath(path: Path2D, transform?: DOMMatrix2DInit): void { }
    // @ts-ignore - unused param.
    arc(x: number, y: number, radius: number, startAngle: number, endAngle: number, anticlockwise?: boolean): void { }
    // @ts-ignore - unused param.
    arcTo(x1: number, y1: number, x2: number, y2: number, radius: number): void { }
    // @ts-ignore - unused param.
    bezierCurveTo(cp1x: number, cp1y: number, cp2x: number, cp2y: number, x: number, y: number): void { }
    // @ts-ignore - unused param.
    closePath(): void { }
    // @ts-ignore - unused param.
    ellipse(x: number, y: number, radiusX: number, radiusY: number, rotation: number, startAngle: number, endAngle: number, anticlockwise?: boolean): void { }
    // @ts-ignore - unused param.
    lineTo(x: number, y: number): void { }
    // @ts-ignore - unused param.
    moveTo(x: number, y: number): void { }
    // @ts-ignore - unused param.
    quadraticCurveTo(cpx: number, cpy: number, x: number, y: number): void { }
    // @ts-ignore - unused param.
    rect(x: number, y: number, w: number, h: number): void { }
}
