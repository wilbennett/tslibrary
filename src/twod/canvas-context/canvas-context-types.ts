import { Color } from '../../colors';


export type CanvasColor = string | Color;
export type Brush = CanvasColor | CanvasGradient | CanvasPattern;

export const enum Composition {
    sourceOver = "source-over",
    sourceIn = "source-in",
    sourceOut = "source-out",
    sourceAtop = "source-atop",
    destinationOver = "destination-over",
    destinationIn = "destination-in",
    destinationOut = "destination-out",
    destinationAtop = "destination-atop",
    lighter = "lighter",
    copy = "copy",
    xor = "xor",
    multiply = "multiply",
    screen = "screen",
    overlay = "overlay",
    darken = "darken",
    lighten = "lighten",
    colorDodge = "color-dodge",
    colorBurn = "color-burn",
    hardLight = "hard-light",
    softLight = "soft-light",
    difference = "difference",
    exclusion = "exclusion",
    hue = "hue",
    saturation = "saturation",
    color = "color",
    luminosity = "luminosity"
}

export type Compositions =
    | "source-over"
    | "source-in"
    | "source-out"
    | "source-atop"
    | "destination-over"
    | "destination-in"
    | "destination-out"
    | "destination-atop"
    | "lighter"
    | "copy"
    | "xor"
    | "multiply"
    | "screen"
    | "overlay"
    | "darken"
    | "lighten"
    | "color-dodge"
    | "color-burn"
    | "hard-light"
    | "soft-light"
    | "difference"
    | "exclusion"
    | "hue"
    | "saturation"
    | "color"
    | "luminosity";
