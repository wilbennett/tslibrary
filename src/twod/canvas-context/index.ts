export * from "./canvas-context-types";
export * from "./canvas-context";
