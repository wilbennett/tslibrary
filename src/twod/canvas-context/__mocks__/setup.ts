import { ImageData } from './image-data';
import { Path2D } from './path2d';

declare global {
    interface Window {
        ImageData: any;// ImageData; // TODO: Typescript complains about missing properties.
        Path2D: any;
    }
}

window.ImageData = ImageData;
window.Path2D = Path2D;
