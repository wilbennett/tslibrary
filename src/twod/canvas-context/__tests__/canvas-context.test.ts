import { MathEx } from '../../../core';
import { PVector2, Vector2 } from '../../../vectors';
import { CanvasRenderingContext2D } from '../__mocks__/canvas-rendering-context2d';
import { CanvasContext } from '../canvas-context';

let shouldReturnNullContext = false;

// @ts-ignore - unused param.
HTMLCanvasElement.prototype.getContext = function (contextId: "2d", ...rest: any[]) {
    if (shouldReturnNullContext) {
        shouldReturnNullContext = false;
        return null;
    }

    return new CanvasRenderingContext2D();
}

test("Create with context", () => {
    const canvas = document.createElement("canvas");
    const ctx = canvas.getContext("2d");

    expect(() => { new CanvasContext(ctx!) }).not.toThrow();
});

it("Should throw error when canvas context is null", () => {
    shouldReturnNullContext = true;
    const canvas = document.createElement("canvas");

    expect(() => { new CanvasContext(canvas) }).toThrow();
});

let context: CanvasContext;
let ctx: CanvasRenderingContext2D;

beforeEach(() => {
    const canvas = document.createElement("canvas");
    context = new CanvasContext(canvas);
    // @ts-ignore - missing members.
    ctx = context.ctx;
});

describe("Should handle canvas transform operations", () => {
    it("Should be identity on creation", () => {
        const identity = context.createIdentity().toString();

        expect(ctx.transformation.toString()).toBe(identity);
        expect(context.getTransform(context.createValues()).toString()).toBe(identity);
        expect(context.getIdentity([]).toString()).toBe(identity);
        expect(context.inverse.toString()).toBe(identity);
        expect(context.getInverse([]).toString()).toBe(identity);
    });

    it("Should handle updating transform", () => {
        const identity = context.createIdentity().toString();
        const translate = [1, 0, 0, 1, 10, 20];
        const scale = [2, 0, 0, 4, 0, 0];
        const expected = "2,0,0,4,10,20";

        context.transform(translate);
        context.transform(scale);
        expect(ctx.transformation.toString()).toBe(expected);

        context.setToIdentity();
        expect(ctx.transformation.toString()).toBe(identity);
        context.transform(translate[0], translate[1], translate[2], translate[3], translate[4], translate[5]);
        context.transform(scale[0], scale[1], scale[2], scale[3], scale[4], scale[5]);
        expect(ctx.transformation.toString()).toBe(expected);
    });

    it("Should handle pushing and popping", () => {
        const identity = context.createIdentity().toString();
        const trans = [1, 0, 0, 1, 10, 20];
        const scale = [2, 0, 0, 4, 0, 0];
        const scaleTrans = [2, 0, 0, 4, 10, 20].toString();

        expect(() => context.popTransform()).toThrow();
        context.pushTransform();
        context.setTransform(trans);
        expect(ctx.transformation.toString()).toBe(trans.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(identity);
        expect(() => context.popTransform()).toThrow();

        context.setTransform(trans);
        context.pushTransform();
        context.setToIdentity();
        expect(ctx.transformation.toString()).toBe(identity);
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());

        context.setToIdentity();
        context.pushTransform(trans);
        expect(ctx.transformation.toString()).toBe(identity);
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());

        context.setTransform(trans);
        context.pushTransform();
        context.resetTransform();
        expect(() => context.popTransform()).toThrow();

        context.setTransform(trans);
        context.pushThenIdentity();
        expect(ctx.transformation.toString()).toBe(identity);
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());

        context.setToIdentity();
        context.setTransform(trans);
        context.pushTransform();
        context.setTransform(scale);
        context.pushTransform();
        expect(ctx.transformation.toString()).toBe(scale.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(scale.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());
        expect(() => context.popTransform()).toThrow();

        context.setToIdentity();
        context.setTransform(trans);
        context.pushThenSet(scale);
        expect(ctx.transformation.toString()).toBe(scale.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());

        context.setToIdentity();
        context.setTransform(trans);
        context.pushThenUpdate(scale);
        expect(ctx.transformation.toString()).toBe(scaleTrans.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());

        context.setToIdentity();
        context.setTransform(trans);
        context.setThenPush(scale);
        expect(ctx.transformation.toString()).toBe(scale.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(scale.toString());

        context.setToIdentity();
        context.setTransform(trans);
        context.updateThenPush(scale);
        expect(ctx.transformation.toString()).toBe(scaleTrans.toString());
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(scaleTrans.toString());

        context.setToIdentity();
        context.pushTransform(scale);
        context.pushTransform(trans);
        context.popTransform();
        expect(ctx.transformation.toString()).toBe(trans.toString());
        context.popUpdate();
        expect(ctx.transformation.toString()).toBe(scaleTrans.toString());
    });

    it("should handle translation", () => {
        const original = new PVector2(1, 0);
        const trans = Vector2.create(10, 10);
        const expected = new PVector2(original.x + trans.x, original.y + trans.y);

        context.setTranslation(trans);
        let transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setTranslation(trans.x, trans.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.translate(trans);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.translate(trans.x, trans.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();
    });

    it("should handle rotation", () => {
        const original = new PVector2(1, 0);
        const angle = 90;
        const center = Vector2.create(10, 10);
        const expected = new PVector2(0, 1);
        const expectedCentered = new PVector2(20, 1);

        context.setRotation(MathEx.toRadians(angle));
        let transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setRotationDegrees(angle);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.rotate(MathEx.toRadians(angle));
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.rotate(MathEx.toRadians(angle), center);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expectedCentered)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.rotate(MathEx.toRadians(angle), center.x, center.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expectedCentered)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.rotateDegrees(angle);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.rotateDegrees(angle, center);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expectedCentered)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.rotateDegrees(angle, center.x, center.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expectedCentered)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();
    });

    it("should handle skewing", () => {
        const original = new PVector2(10, 10);
        const angle = Vector2.create(10, 20);
        const radians = new PVector2(MathEx.toRadians(angle.x), MathEx.toRadians(angle.y));
        const expected = new PVector2(11.76, 13.64);

        context.setSkew(radians);
        let transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setSkew(radians.x, radians.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.setSkewDegrees(angle);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setSkewDegrees(angle.x, angle.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.skew(radians);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skew(radians.x, radians.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewX(radians.x);
        transformed = context.transformPoint(original);
        expect(transformed.x).toBeCloseTo(expected.x);
        expect(transformed.y).toBe(original.y);
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewY(radians.y);
        transformed = context.transformPoint(original);
        expect(transformed.y).toBeCloseTo(expected.y);
        expect(transformed.x).toBe(original.x);
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewDegrees(angle);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewDegrees(angle.x, angle.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected, 0.01)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewXDegrees(angle.x);
        transformed = context.transformPoint(original);
        expect(transformed.x).toBeCloseTo(expected.x);
        expect(transformed.y).toBe(original.y);
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.skewYDegrees(angle.y);
        transformed = context.transformPoint(original);
        expect(transformed.y).toBeCloseTo(expected.y);
        expect(transformed.x).toBe(original.x);
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();
    });

    it("should handle scaling", () => {
        const original = new PVector2(1, 2);
        const scale = Vector2.create(2, 4);
        let expected = new PVector2(original.x * scale.x, original.y * scale.y);

        context.setScale(scale);
        let transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setScale(scale.x, scale.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.scale(scale);
        transformed = context.transformPoint(original);//? $.toString()
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.setToIdentity();
        context.scale(scale.x, scale.y);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        expected = new PVector2(original.x * scale.x, original.y * scale.x);
        context.setToIdentity();
        context.setScale(scale.x);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();

        context.scale(scale.x);
        transformed = context.transformPoint(original);
        expect(transformed.equals(expected)).toBeTruthy();
        expect(context.transformPointInverse(transformed).equals(original)).toBeTruthy();
    });
});

describe("Should handle pushing and popping", () => {
    it("Should handle pushing and popping properties", () => {
        expect(() => context.popProps()).toThrow();
        context.pushProps();
        context.popProps();
        expect(() => context.popProps()).toThrow();
    });

    it("Should handle pushing and popping properties and transforms", () => {
        const identity = context.createIdentity().toString();
        const initial = [1, 1, 1, 1, 1, 1];

        expect(() => context.restore()).toThrow();
        context.save();
        context.setTransform(initial[0], initial[1], initial[2], initial[3], initial[4], initial[5]);
        expect(ctx.transformation.toString()).toBe(initial.toString());
        context.restore();
        expect(ctx.transformation.toString()).toBe(identity);
        expect(() => context.restore()).toThrow();
    });
});
