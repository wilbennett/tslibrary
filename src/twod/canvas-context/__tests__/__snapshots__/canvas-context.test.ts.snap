// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle pushing and popping Should handle pushing and popping properties 1`] = `
Object {
  "fillStyle": "red",
  "font": "20px Arial",
  "globalAlpha": 0.5,
  "globalCompositeOperation": "source-in",
  "imageSmoothingEnabled": true,
  "imageSmoothingQuality": "high",
  "lineCap": "round",
  "lineDashOffset": 2,
  "lineJoin": "bevel",
  "lineWidth": 3,
  "miterLimit": 4,
  "shadowBlur": 0.6,
  "shadowColor": "blue",
  "shadowOffsetX": 10,
  "shadowOffsetY": 20,
  "strokeStyle": "green",
  "textAlign": "center",
  "textBaseline": "bottom",
}
`;
